[Skia][GStreamer] Implement ImageGStreamer

carlosgcampos opened the pull request:
#### 5afce8e3f3a69057b9eb565af0c341d72d23ce62
<pre>
[Skia][GStreamer] Implement ImageGStreamer
<a href="https://bugs.webkit.org/show_bug.cgi?id=269907">https://bugs.webkit.org/show_bug.cgi?id=269907</a>

Reviewed by Philippe Normand.

* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
(WebCore::ImageGStreamer::ImageGStreamer):
(WebCore::ImageGStreamer::~ImageGStreamer):

Canonical link: <a href="https://commits.webkit.org/275224@main">https://commits.webkit.org/275224@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c76b0e09de60eddde0a62351c5ba2bb9328285aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41104 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20117 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43666 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43411 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23230 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17448 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34034 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41678 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17109 "Found 1 new test failure: http/wpt/background-fetch/background-fetch-persistency.window.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35407 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14667 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14846 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36407 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44981 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37361 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36721 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40473 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15919 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13056 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38849 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17538 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17590 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5502 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17182 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
```diff
@@ -23,23 +23,98 @@
+    if (UNLIKELY(!GST_IS_BUFFER(buffer)))
+        return;
+
+    auto* videoFrame = static_cast<GstVideoFrame*>(fastMalloc(sizeof(GstVideoFrame)));
```
`GstMappedFrame` could be used here instead.

carlosgcampos committed:
[Skia][GStreamer] Implement ImageGStreamer
https://bugs.webkit.org/show_bug.cgi?id=269907

Reviewed by Philippe Normand.

* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
(WebCore::ImageGStreamer::ImageGStreamer):
(WebCore::ImageGStreamer::~ImageGStreamer):

Canonical link: https://commits.webkit.org/275224@main

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -272,21 +273,31 @@ class GstMappedFrame {
+        return m_isValid ? GST_VIDEO_FRAME_PLANE_DATA(&m_frame, planeIndex) : nullptr;
+    }
+
+    int planeStride(uint32_t planeIndex) const
+    {
+        return m_isValid ? GST_VIDEO_FRAME_PLANE_STRIDE(&m_frame, planeIndex) : -1;
```
I don't have a strong opinion about this but I would have made a different call here. I understand -1 is the "no stride option" for skia but in the case of the API we are defining here I would make this as `std::optional<int> planeStride...` and when calling this use `planeStride(x).value_or(-1)`

carlosgcampos replied:
No, this has nothing to do with Skia, this will never be used from Skia when it's invalid. I just followed what the other methods do for consistency. I think we shouldn't handle the invalid case, it's a programmer error to call any method on an invalid mapped frame, so I would just ASSERT(m_isValid) instead in all the methods.
